### PR TITLE
fix ncbi id issue

### PR DESCRIPTION
**Pull request by blubbundbla**
This should fix the issue we had with the ids in the test runs.
I added functions to get the Genbank accession numbers and Genbank identifier (if available) and rewrote parts of read_local_blast_query() to not use web-queries.

### CODE REVIEW

File: physcraper/__init__.py [hunk 1]
```diff
@@ -1812,6 +1809,44 @@ def run_blast_wrapper(self):  # TODO Should this be happening elsewhere?
                                                                                                             last_blast))
         self._blasted = 1
 
+    def get_acc_from_blast(self, query_string):
```
Comment (snacktavish) [inline physcraper/__init__.py]: Will these only work with a local blast database? If so would be good to make that clear in function name - (e.g. get_acc_from_local_blast). 
Is there any way to test within this function that the value is actually an accession number? (e.g. in case there is an error in the query string that is fed in). Or to test that the query string has some specific characteristics?

File: physcraper/__init__.py [hunk 2]
```diff
@@ -1834,26 +1869,30 @@ def read_local_blast_query(self, fn_path):
...
-                # Get all of them! (get redundant seq info)
+                # Get all of them! (they can be of a different taxon ids = get redundant seq info)
                 found_taxids = set()
                 found_spn = set()
-                if len(sallseqid.split(";")) > 1:
-                    # get additional info only for seq that pass the eval
-                    if evalue < float(self.config.e_value_thresh):
+                # get additional info only for seq that pass the eval
+                if evalue < float(self.config.e_value_thresh):
+                    if len(sallseqid.split(";")) > 1:
```
Comment (snacktavish) [inline physcraper/__init__.py]: Pulling out this search for merged tax_ids into as separate function would make this section easier to read and debug. If I'm understanding correctly, this if statement is only true if there is more that one id for the sequence? It is hard to find where the 'else' to this 'if' is, which is a sign that this whole block should be pulled out into a seperate function.

File: physcraper/__init__.py [hunk 3]
```diff
@@ -1981,6 +2017,39 @@ def read_local_blast_query(self, fn_path):
...
+        if not self.config.blastdb == "./tests/data/precooked/testing_localdb":
+            os.system(cmd1)
+            f = open("{}/tmp/tax_id_{}.csv".format(self.workdir, gb_acc))
+            tax_id_l = []
+            for line in iter(f):
+                line = line.rstrip().lstrip()
+                tax_id_l.append(int(line))
+            f.close() 
+        else: 
+            f = open("./tests/data/precooked/testing_localdb/tax_id_{}.csv".format(gb_acc))
```
Comment (snacktavish) [inline physcraper/__init__.py]: I'm not sure what exactly is happening here, but I don't think we want to hardcode in something from testing as the blast database.

File: physcraper/__init__.py [hunk 4]
```diff
@@ -3256,16 +3323,14 @@ def loop_for_write_blast_files(self, tax_id):
         """
         debug("loop_for_write_blast_files")
         aln_otus = set([taxon.label for taxon in self.data.aln])
-        for otu_id in self.sp_d[tax_id]:
+        query_otu = None
+        db_otus = []
```
Comment (snacktavish) [inline physcraper/__init__.py]: I think this can be deleted? I don't see it being used.

File: physcraper/__init__.py [hunk 5]
```diff
@@ -1981,6 +2017,39 @@ def read_local_blast_query(self, fn_path):
                                                query_dict[key]["evalue"]))
                 fn.close()
 
+    def get_taxid_from_acc(self, gb_acc):
```
Comment (snacktavish) [inline physcraper/__init__.py]: This is only for local blast, right? Maybe put in name, and or add an assert right at the beginning of the function that the configuration is for local blast.

File: physcraper/__init__.py (hunk 1 above)
Comment (blubbundbla) [reply physcraper/__init__.py]: No, looks like it also works for web blasting. Adjusted doc string accordingly.
Yeah, not sure how else to test, that the right string was fed in.

File: physcraper/__init__.py (hunk 3 above)
Comment (blubbundbla) [reply physcraper/__init__.py]: I implemented this so that we can still run tests. This function relies on the blast db and as we do not have them, we need to get the stored files from the fprecooked folder.

File: physcraper/__init__.py (hunk 5 above)
Comment (blubbundbla) [reply physcraper/__init__.py]: see above

File: physcraper/__init__.py (hunk 1 above)
Comment (snacktavish) [reply physcraper/__init__.py]: Sounds good.

File: physcraper/__init__.py (hunk 3 above)
Comment (snacktavish) [reply physcraper/__init__.py]: Ideally in testing we are running the same commands as when it isn't for testing. It looks like it is creating a csv for each accession number with taxon ids that go with that accession number, and then looks for them in workdir/tmp. If that is correct, one way to make it work the same for the tests is to copy the pre-cooked 'tax_id_gb_acc.csv' into tmp/workdir as part of the test, rather than hardcoding it into the code. Do you think that will work?

File: physcraper/__init__.py [hunk 6]
```diff
@@ -1809,186 +1810,168 @@ def run_blast_wrapper(self):  # TODO Should this be happening elsewhere?
                                                                                                             last_blast))
         self._blasted = 1
 
-    def get_acc_from_blast(self, query_string):
+    def get_new_seqs_for_mergedseq(self, var_list, query_dict):
```
Comment (snacktavish) [inline physcraper/__init__.py]: This is an improvement! But it still has a pretty hard to follow the long series of if-else's. Is it possible to separate out the different pieces into functions, rather than all together in one? e.g. a function that takes each  staxids_l (I'm still not clear on what the staxids_l has in in it...) and returns the info you need for it? Or is there another place you can find that it would be possible to separate this into more pieces?

File: physcraper/__init__.py [hunk 7]
```diff
@@ -1809,186 +1810,168 @@ def run_blast_wrapper(self):  # TODO Should this be happening elsewhere?
...
+            except AssertionError:
+                sys.stderr.write("Genbank accession {} does map to different taxonid in database: {} vs {}. Check with ncbi\n".format(gb_acc, item, staxids_l)) 
+
+        # this while loop is here to speed up the process of finding the correct information
+        count = 0
+        spn_range = 0
+        stop_while = False
+        id_before = 0
+        id_now = -2
+        while len(found_taxids) < len(set(staxids_l)):  # as long as i have not found all taxids for the seq
```
Comment (snacktavish) [inline physcraper/__init__.py]: Is there a reason you're not just iterating through the item in staxids_l?

File: physcraper/__init__.py [hunk 8]
```diff
@@ -1809,186 +1810,168 @@ def run_blast_wrapper(self):  # TODO Should this be happening elsewhere?
...
+                    stitle = salltitles_l[i]
+                    # if both var are the same, we do not need to search GB for taxon info
+                    # staxids = tax_id_l[i]
+                    print("check out what goes wrong here")
+                    print(gb_acc, self.get_taxid_from_acc(gb_acc))
+                    print(gb_acc, self.get_taxid_from_acc(gb_acc)[0])
+
+                    staxids = int(self.get_taxid_from_acc(gb_acc)[0])  # corresponding taxid is always in first position
+                    # debug(qtaxid)
+                    id_before = id_now
```
Comment (snacktavish) [inline physcraper/__init__.py]: I don't see these being used? Can they be deleted?

File: physcraper/__init__.py [hunk 9]
```diff
@@ -1809,186 +1810,168 @@ def run_blast_wrapper(self):  # TODO Should this be happening elsewhere?
...
+          
+                    for item in tax_id_l:
+                        try:
+                            assert str(staxids) in staxids_l or str(staxids) in tax_id_l, (item, staxids_l, tax_id_l, gb_acc)
+                        except AssertionError:
+                            sys.stderr.write("Merged Genbank accession {} does map to different taxonid in database: {} vs {}. Check with ncbi\n".format(gb_acc, item, staxids_l)) 
+                    # assert str(staxids) in staxids_l or staxids in tax_id_l, (str(staxids), staxids_l, tax_id_l)
+                    
+                    # if gb acc was already read in before stop the for loop
+                    if gb_acc in query_dict or gb_acc in self.data.gb_dict:
```
Comment (snacktavish) [inline physcraper/__init__.py]: For example, if you have a function that returns a gb_acc and a taxon_id (or whatever other information is required) for each sequence id, and separate function  adds that information into the query dict for each accession number? That would have the advantage of being a function that could be reused for adding information into the query dict for unmerged sequences as well.

File: physcraper/__init__.py [hunk 10]
```diff
@@ -1809,186 +1810,168 @@ def run_blast_wrapper(self):  # TODO Should this be happening elsewhere?
...
+                                {'^ncbi:gi': gi_id, 'accession': gb_acc, 'staxids': staxids,
+                                 'sscinames': sscinames, 'pident': pident, 'evalue': evalue,
+                                 'bitscore': bitscore, 'sseq': sseq, 'title': stitle}
+                    except AssertionError:
+                        sys.stderr.write("Taxon Id {} was not initally found in the blast search - corresponding Genbank accession number is: {}. Sequence will not be added. Check with ncbi. \n".format(staxids, gb_acc))
+         
+            # same loop as above, only that it does more blastdbcmd's
+            # this is used as sometimes the if above does not yield in stop_while == True,
+            # through different taxa names
+            elif count >= 1 and stop_while is False:
```
Comment (snacktavish) [inline physcraper/__init__.py]: I see the comment above - but why would the original loop not get all the taxon names? Is it posible to moidify it, so that we don't need to come back to it here?

**Comment by snacktavish** [comment]
Incorporated into refactor branch
